INT-4544: Allow runtime MBeans (un)registered

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4544

* Fix `SourcePollingChannelAdapterFactoryBean` to register an
`outputChannelName` into the `SourcePollingChannelAdapter` for late
binding, especially in case of dynamic `IntegrationFlow` registrations.
This must be back-ported to `5.0.x`
* Implement `DestructionAwareBeanPostProcessor` in the
`IntegrationMBeanExporter` for the runtime beans tracking, e.g. via
dynamic `IntegrationFlow` registrations
* Refactor `IntegrationMBeanExporter` for static and dynamic beans
registration and destruction
* Remove unused properties; introduce some new for tracking beans and
their relationship

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
INT-4544: Allow runtime MBeans (un)registered

JIRA: https://jira.spring.io/browse/INT-4544

* Fix `SourcePollingChannelAdapterFactoryBean` to register an
`outputChannelName` into the `SourcePollingChannelAdapter` for late
binding, especially in case of dynamic `IntegrationFlow` registrations.
This must be back-ported to `5.0.x`
* Implement `DestructionAwareBeanPostProcessor` in the
`IntegrationMBeanExporter` for the runtime beans tracking, e.g. via
dynamic `IntegrationFlow` registrations
* Refactor `IntegrationMBeanExporter` for static and dynamic beans
registration and destruction
* Remove unused properties; introduce some new for tracking beans and
their relationship

garyrussell commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -300,6 +306,93 @@ public void afterSingletonsInstantiated() {
 
 	}
 
+	private void registerProducer(MessageProducer messageProducer) {
```
Should be below the next method.

artembilan replied:
It is called from the `afterSingletonsInstantiated()` first time, so this is normal order for the private methods.

garyrussell replied:
But it's also called from `postProcessAfterInitialization()`. I believe the rule is called methods should always be below the last place that calls it.

artembilan replied:
Fine with me, but I've followed this rule so far: https://github.com/spring-projects/spring-framework/wiki/Code-Style#java-source-file-organization

> Note that private or protected methods called from method implementations should be placed immediately below the methods where they're used. In other words if there 3 interface method implementations with 3 private methods (one used from each), then the order of methods should include 1 interface and 1 private method in sequence, not 3 interface and then 3 private methods at the bottom.

garyrussell replied:
Maybe it's just my interpretation then; I guess those guidelines don't specifically address this use case.

artembilan replied:
OK. Fine with me.
Will fix. 😄 

Not sure though, what happened to the `org.springframework.integration.monitor.MessageSourceTests` on Travis...

garyrussell reviewed (changes requested):
That's all after a quick review - will perform a more thorough review tomorrow (after travis).

artembilan committed:
* Move `IntegrationMBeanExporter.registerProducer()` after
`postProcessAfterInitialization()`
* Refactor JMX test configurations to reuse MBeanServer as much as
possible and destroy the server whenever it is necessary

artembilan committed:
* More JMX tests polishing

garyrussell commented:
Looks like there are still test issues (instance already exists exceptions).

artembilan commented:
Yeah... Looks like MBean Server is not closed somewhere yet...

artembilan committed:
* And more JMX tests polishing

garyrussell commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -292,12 +296,100 @@ public void afterSingletonsInstantiated() {
+
+	@Override
+	public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+		if (this.singletonsInstantiated) {
```
Since we are already at RC, perhaps we should add a `try {...} catch {...}` under here to fail soft (and log an error) instead of failing hard (just in case there are any regression issues) ?

garyrussell reviewed (changes requested):
And still travis failure.

artembilan committed:
* `try...catch` in the `IntegrationMBeanExporter.postProcessAfterInitialization()`
to avoid breaking changes for the current GA phase
* Do not unregister those MBeans explicitly which weren't created at
runtime

artembilan committed:
* Log ObjectNames in the `MBeanExporterIntegrationTests`

artembilan committed:
* Fix `MessageMetricsAdviceTests` to rely on the application context
* Do not use a cst in the `StandardIntegrationFlowContext` for
`BeanFactory`, but an explicit `BeanDefinitionRegistry`
* Extract targets from proxies in the
`IntegrationMBeanExporter.postProcessAfterInitialization()`
* Remove logging in the `MBeanExporterIntegrationTests`

artembilan committed:
* Fix `NotificationListeningMessageProducerTests` to reuse existing `MBeanServer`

artembilan committed:
* Change JMX `domain` in the `DslMBeanTests` to `dsl` do not clash with
similar in the `foo` in the `MessageSourceTests`
* Use `MBeanServer` bean in the `Int2307Tests`
* Use JUnit 5ctor injection injection in the `MessageMetricsAdviceTests`

artembilan committed:
* Fix JMX tests do not reuse existing `MBeanServer` and make them rely
on the server provided by the managed `MBeanServerFactoryBean` which
destroys a server on application context close
* Fix `StoredProcJmxManagedBeanTests-context.xml` to use `MBeanServer`
from the `<context:mbean-server>`

artembilan commented:
OK! Finally it is green 😄 

Sorry for the mess in tests.

Note: When we configure runtime beans, we don't process plain `MessageHandler` and `MessageSource` beans. I would say without endpoints it does not make sense to have MBeans for plain beans.
Plus there is the problem that `postProcessAfterInitialization()` is called for `MessageHandler`/`MessageSource` and their endpoint, but at the moment to parse one of them there is no another yet...

garyrussell commented:
> This must be back-ported to 5.0.x

Separate PR?
